kola/tests/kubeadm/templates.go

```diff
@@ -0,0 +1,205 @@
+        sed "s:/usr/bin:${DOWNLOAD_DIR}:g" |
+        tee /etc/systemd/system/kubelet.service
+    
+    curl -sSL "https://raw.githubusercontent.com/kubernetes/release/${RELEASE_VERSION}/cmd/kubepkg/templates/latest/deb/kubeadm/10-kubeadm.conf" |
+        sed "s:/usr/bin:${DOWNLOAD_DIR}:g" |
+        tee /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+    
+    curl -sSL --remote-name-all https://storage.googleapis.com/kubernetes-release/release/${RELEASE}/bin/linux/amd64/{kubeadm,kubelet,kubectl}
+    
+    chmod +x {kubeadm,kubelet,kubectl}
+    mv {kubeadm,kubelet,kubectl} "${DOWNLOAD_DIR}"
+    
+    systemctl enable --now kubelet
+   
+    # we create the kubeadm config
+    # plugin-volume-dir and flex-volume-plugin-dir are required since /usr is read-only mounted
+    # etcd is also defined as external. The provided one has some issues with docker and selinux
+    # (permission denied with /var/lib/etcd) so it can't boot properly
+    cat << EOF > kubeadm-config.yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: InitConfiguration
+nodeRegistration:
+  kubeletExtraArgs:
+    volume-plugin-dir: "/opt/libexec/kubernetes/kubelet-plugins/volume/exec/"
+---
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+networking:
+  podSubnet: {{ .PodSubnet }}
+controllerManager:
+  extraArgs:
+    flex-volume-plugin-dir: "/opt/libexec/kubernetes/kubelet-plugins/volume/exec/"
+etcd:
+  external:
+    endpoints:
+    {{ range $endpoint := .Endpoints }}
+      - {{ $endpoint }}
+    {{ end }}
+EOF
+    
+    kubeadm config images pull
+    kubeadm init --config kubeadm-config.yaml
+    
+    cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
+    cp -i /etc/kubernetes/admin.conf /home/core/.kube/config
+    sudo chown -R core:core /home/core/.kube; chmod a+x /home/core/.kube/config;
+    
+    cat << EOF > calico.yaml
+# Source: https://docs.projectcalico.org/manifests/custom-resources.yaml
+apiVersion: operator.tigera.io/v1
+kind: Installation
+metadata:
+  name: default
+spec:
+  # Configures Calico networking.
+  calicoNetwork:
+    # Note: The ipPools section cannot be modified post-install.
+    ipPools:
+      - blockSize: 26
+        cidr: {{ .PodSubnet }}
+        encapsulation: VXLANCrossSubnet
+        natOutgoing: Enabled
+        nodeSelector: all()
+  flexVolumePath: /opt/libexec/kubernetes/kubelet-plugins/volume/exec/
+EOF
+    
+    kubectl create -f https://docs.projectcalico.org/manifests/tigera-operator.yaml
+    kubectl apply -f calico.yaml
+} > /dev/null 2>&1
+
+URL=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}')
+prefix="https://"
+short_url=${URL#"${prefix}"}
+
+cat << EOF
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: JoinConfiguration
+discovery:
+  bootstrapToken:
+    apiServerEndpoint: ${short_url}
+    token: $(kubeadm token create)
```

Comment:
errors of this subshell won't be caught, it's better to first run this line in a single assignment to a variable and then use this variable here